vm/starnix/starnix.go

```diff
@@ -0,0 +1,383 @@
+}
+
+func (inst *instance) terminateFuchsiaVM() {
+	if inst.debug {
+		log.Logf(0, "terminating Fuchsia VM: %v", inst.name)
+	}
+	inst.runCommand(1*time.Minute*inst.timeouts.Scale, inst.ffxBinary, "emu", "stop", inst.name)
+}
+
+func (inst *instance) Close() {
+	if inst.fuchsiaVMStarted {
+		inst.terminateFuchsiaVM()
+	}
+	if inst.fuchsiaLogs != nil {
+		inst.fuchsiaLogs.Process.Kill()
+		inst.fuchsiaLogs.Wait()
+	}
+	if inst.adb != nil {
+		inst.adb.Process.Kill()
+		inst.adb.Wait()
+	}
+	if inst.merger != nil {
+		inst.merger.Wait()
+	}
+	if inst.rpipe != nil {
+		inst.rpipe.Close()
+	}
+	if inst.wpipe != nil {
+		inst.wpipe.Close()
+	}
+}
+
+func (inst *instance) startFuchsiaVM() error {
+	ffxArgs := []string{"emu", "start", "--headless", "--name", inst.name}
+	inst.args = ffxArgs[2:]
+	err := inst.runCommand(5*time.Minute*inst.timeouts.Scale, inst.ffxBinary, ffxArgs...)
+	if err != nil {
+		return err
+	}
+	inst.fuchsiaVMStarted = true
+	return nil
+}
+
+func (inst *instance) startFuchsiaLogs() (*exec.Cmd, error) {
+	cmd := osutil.Command(inst.ffxBinary, "--target", inst.name, "log")
+	cmd.Dir = inst.fuchsiaDirectory
+	cmd.Stdout = inst.wpipe
+	cmd.Stderr = inst.wpipe
+	inst.merger.Add("fuchsia", inst.rpipe)
+	return cmd, cmd.Start()
+}
+
+func (inst *instance) startAdbServerAndConnection(timeout time.Duration) error {
+	cmd := osutil.Command(inst.ffxBinary, "--target", inst.name, "starnix", "adb", "-p", fmt.Sprintf("%d", inst.port))
+	cmd.Dir = inst.fuchsiaDirectory
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	if inst.debug {
+		log.Logf(0, fmt.Sprintf("the adb bridge is listening on 127.0.0.1:%d", inst.port))
+	}
+	inst.adb = cmd
+	return inst.connectToAdb(timeout)
+}
+
+func (inst *instance) connectToAdb(timeout time.Duration) error {
+	startTime := time.Now()
+	for {
+		vmimpl.SleepInterruptible(3 * time.Second)
+		if inst.debug {
+			log.Logf(0, "attempting to connect to ADB")
+		}
+		cmd := osutil.Command("adb", "connect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		connectOutput, err := cmd.Output()
+		if err == nil && strings.HasPrefix(string(connectOutput), "connected to") {
+			return nil
+		}
+		cmd = osutil.Command("adb", "disconnect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		cmd.Run()
+		if inst.debug {
+			log.Logf(0, "adb connect failed")
+		}
+		if time.Since(startTime) > timeout {
+			return fmt.Errorf("can't connect to ADB server")
+		}
+	}
+}
+
+// Script for telling syz-fuzzer how to connect to syz-executor.
+func (inst *instance) createAdbScript() error {
+	adbScript := fmt.Sprintf(
+		`#!/bin/bash
+		adb_port=$1
+		fuzzer_args=${@:2}
+		adb -s 127.0.0.1:$adb_port shell "cd %s; ./syz-executor $fuzzer_args"`, inst.targetDir)
+	return os.WriteFile(inst.executor, []byte(adbScript), 0777)
+}
+
+func (inst *instance) restartDaemon() {
+	if inst.debug {
+		log.Logf(0, "restarting daemon")
+	}
+	osutil.RunCmd(1*time.Minute*inst.timeouts.Scale, inst.fuchsiaDirectory, "ffx", "doctor", "--restart-daemon")
```

Comment:
Is it a different ffx binary from what used in other places?
If not, call is the same way with full path: `inst.runCommand(1*time.Minute*inst.timeouts.Scale, inst.ffxBinary, ...)` and inline.

Since ffx is run lots of times, this can be reduced to just: `inst.ffx("doctor", "--restart-daemon")`.